File: leetcode/AIcoding/4.GPT-4o/cpp_solutions/1000-minimum-cost-to-merge-stones.cpp
class Solution {
public:
    int mergeStones(vector<int>& stones, int K) {
        int n = stones.size();
        if ((n - 1) % (K - 1) != 0) return -1;

        vector<vector<int>> dp(n + 1, vector<int>(n + 1, INT_MAX));
        vector<vector<int>> prefix(n + 1, vector<int>(n + 1, 0));

        for (int i = 1; i <= n; ++i) {
            prefix[i][i] = stones[i - 1];
            for (int j = i + 1; j <= n; ++j) {
                prefix[i][j] = prefix[i][j - 1] + stones[j - 1];
            }
        }

        for (int len = 1; len <= n; ++len) {
            for (int i = 1; i <= n - len + 1; ++i) {
                int j = i + len - 1;
                if (len == 1) {
                    dp[i][j] = 0;
                } else {
                    for (int k = 1; k < len; ++k) {
                        dp[i][j] = min(dp[i][j], dp[i][i + k - 1] + dp[i + k][j]);
                    }
                    if ((len - 1) % (K - 1) == 0) {
                        dp[i][j] += prefix[i][j];
                    }
                }
            }
        }

        return dp[1][n];
    }
};

File: leetcode/AIcoding/4.GPT-4o/python3_solutions/1000-minimum-cost-to-merge-stones.py
class Solution:
    def mergeStones(self, stones: List[int], K: int) -> int:
        n = len(stones)
        if (n - 1) % (K - 1) != 0:
            return -1

        dp = [[[float('inf')] * (K + 1) for _ in range(n)] for _ in range(n)]
        prefix = [0] * (n + 1)

        for i in range(n):
            prefix[i + 1] = prefix[i] + stones[i]
            dp[i][i][1] = 0

        for length in range(2, n + 1):
            for i in range(n - length + 1):
                j = i + length - 1
                for m in range(1, K + 1):
                    for mid in range(i, j):
                        dp[i][j][m] = min(dp[i][j][m], dp[i][mid][m] + dp[mid + 1][j][m])
                    if m > 1:
                        dp[i][j][1] = min(dp[i][j][1], dp[i][j][m] + prefix[j + 1] - prefix[i])

        return dp[0][n - 1][1]